### PR TITLE
fix: registries fields validations

### PR DESCRIPTION
**Pull request by paulonnn**
## Checklist

  - [x] This PR is **not** opened from my fork's `main` branch

  ## What This PR Implements

  Adds validation and UX improvements to the container registry form, ensuring generic (non-ECR) registries require proper credentials and preventing registry type changes after creation.

  Fixes: (UX improvement)

  ## Changes Made

  - **Backend**: Require `username` and `token` when creating a generic container registry, returning `ValidationError` if either is empty
  - **Backend**: Require `username` on generic registry updates (token is optional on update — empty means "keep existing")
  - **Backend**: Block registry type changes after creation (`generic` ↔ `ecr`), returning `ValidationError` if attempted
  - **Frontend**: Add matching Zod validation in the container registry form (`username` always required, `token` required on create only)
  - **Frontend**: Disable the registry type selector when editing an existing registry (greyed out, non-clickable)
  - **i18n**: Add `registries_username_required` and `registries_token_required` message keys to `en.json`

  ## Testing Done

  - [x] Development environment started: `./scripts/development/dev.sh start`
  - [x] Frontend verified at http://localhost:3000
  - [x] Backend verified at http://localhost:3552
  - [x] Manual testing completed (describe):
    - Verified creating a generic registry without username/token is rejected
    - Verified editing a registry keeps the existing token when the field is left empty
    - Verified changing registry type on update is rejected by the backend
    - Verified the registry type selector is disabled in edit mode on the frontend
  - [x] No linting errors (e.g., `just lint all`)
  - [x] Backend tests pass: `just test backend`

  ## AI Tool Used (if applicable)

  AI Tool: Claude
  Assistance Level: Moderate
  What AI helped with: Explored codebase patterns, planned implementation, wrote validation logic and tests
  I reviewed and edited all AI-generated output: Yes
  I ran all required tests and manually verified changes: Yes

  ## Additional Context

  - `SyncRegistries` is unaffected — it writes directly to the DB and can still change registry types as needed for internal sync
  - Existing registries with empty credentials are not retroactively broken on update; only the create path is fully locked down

**Disclaimer** Greptiles Reviews use AI, make sure to check over its work.

To better help train Greptile on our codebase, if the comment is useful and valid **Like** the comment, if its not helpful or invalid **Dislike**

To have Greptile Re-Review the changes, mention `greptileai`.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds credential validation and UX hardening to the container registry feature: generic registries now require `username` (and `token` on create) at the backend level, registry type changes after creation are blocked, and the frontend mirrors these rules with matching Zod validation and a disabled type selector in edit mode.

**Key changes:**
- `CreateRegistry`: validates non-empty `username` and `token` for generic registries before encrypting/persisting
- `applyRegistryTypeUpdateInternal`: now returns a `ValidationError` when the caller tries to change `RegistryType`, instead of silently migrating credentials
- `updateGenericRegistryFieldsInternal`: adds a post-update username presence check; token remains optional on update (empty = keep existing)
- Frontend `superRefine`: adds matching `username` (always) and `token` (create-only) validation for generic registries
- i18n: two new keys (`registries_username_required`, `registries_token_required`); three existing AWS error strings de-scoped from "for ECR registries" to match the simplified backend messages

**Issue to address:**
- The username validation in `updateGenericRegistryFieldsInternal` checks `registry.Username` *after* applying `req.Username`. When `req.Username` is `nil` (not provided), the field is not updated, so the existing DB value is checked. Any legacy generic registry that was created before this validation existed and has an empty `username` will fail on **any** update operation—including purely non-credential changes like toggling `enabled` or changing `description`. This contradicts the PR description's statement that "Existing registries with empty credentials are not retroactively broken on update." A test covering a description-only update to a legacy registry with a blank username would surface this immediately.
</details>

<details open><summary><h3>Confidence Score: 3/5</h3></summary>

- Safe to merge only if no existing generic registries have empty usernames; otherwise any update to those records will silently start failing.
- The new validations and the frontend changes are well-structured, tests cover the intended happy/sad paths, and the type-lock logic is correct. However, there is a logic gap in `updateGenericRegistryFieldsInternal`: when `req.Username` is nil, the existing (potentially empty) username is validated, meaning legacy registries with blank usernames become un-updatable through any field. This contradicts the stated PR intent and could break existing deployments without a migration or a targeted guard.
- backend/internal/services/container_registry_service.go — specifically the `updateGenericRegistryFieldsInternal` function
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| backend/internal/services/container_registry_service.go | Core service changes: blocks type changes on update and adds credential validation on create/update. Contains a logic issue where any update to a legacy generic registry with an empty username (pre-validation) will now fail, contradicting the stated PR intent. |
| backend/internal/services/container_registry_service_test.go | Good coverage of happy and sad paths for the new validations. Missing a test for the edge case where a legacy registry with empty username is updated via a non-credential field (e.g., description-only change). |
| frontend/src/lib/components/sheets/container-registry-sheet.svelte | Adds matching Zod validation for generic registry credentials and disables the registry-type selector in edit mode. Uses $derived correctly with no $state/$effect anti-patterns. |
| frontend/messages/en.json | Adds two new i18n keys (registries_username_required, registries_token_required) and trims "for ECR registries" from the three existing AWS error strings, aligning with the backend message changes. |

</details>

</details>

<details><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[UpdateRegistry request] --> B[GetRegistryByID]
    B --> C[Update common fields\nurl, description, insecure, enabled]
    C --> D{registryType\nprovided?}
    D -- No --> F
    D -- Yes --> E{type changed?}
    E -- Yes --> ERR1[ValidationError:\ntype cannot change]
    E -- No --> F{RegistryType\nin DB?}
    F -- ECR --> G[updateECRRegistryFieldsInternal]
    F -- Generic --> H[updateGenericRegistryFieldsInternal]
    G --> G1[UpdateIfChanged AWSAccessKeyID / Region]
    G1 --> G2{AWSSecretAccessKey\nprovided?}
    G2 -- Yes --> G3[Encrypt & update secret]
    G2 -- No --> G4
    G3 --> G4[Validate all AWS fields non-empty]
    G4 --> SAVE
    H --> H1[UpdateIfChanged Username]
    H1 --> H2{Token provided\nand non-empty?}
    H2 -- Yes --> H3[Encrypt & update token]
    H2 -- No --> H4
    H3 --> H4{registry.Username\nempty after update?}
    H4 -- Yes --> ERR2[ValidationError:\nusername required]
    H4 -- No --> SAVE
    SAVE[db.Save registry] --> RET[Return updated registry]
```
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: backend/internal/services/container_registry_service.go
Line: 254-269

Comment:
**Legacy registries will fail on any update**

The username validation at line 265 checks `registry.Username` *after* applying the update. If `req.Username` is `nil` (not provided), `UpdateIfChanged` doesn't modify the field, so `registry.Username` retains whatever value is in the DB. For a legacy generic registry that was created before this validation existed (i.e., where `username` is `""`), **any update operation**—even a change to just `url`, `description`, or `enabled`—will now return `"Username is required"`.

This contradicts the PR description's claim that "Existing registries with empty credentials are not retroactively broken on update; only the create path is fully locked down."

A simple guard can preserve the stated intent: only enforce the non-empty constraint when the caller is explicitly touching the `username` field.

```go
func (s *ContainerRegistryService) updateGenericRegistryFieldsInternal(registry *models.ContainerRegistry, req models.UpdateContainerRegistryRequest) error {
	if req.Username != nil {
		if strings.TrimSpace(*req.Username) == "" {
			return &models.ValidationError{Field: "username", Message: "Username is required"}
		}
		utils.UpdateIfChanged(&registry.Username, req.Username)
	}

	if req.Token != nil && *req.Token != "" {
		encryptedToken, err := crypto.Encrypt(*req.Token)
		if err != nil {
			return fmt.Errorf("failed to encrypt token: %w", err)
		}
		utils.UpdateIfChanged(&registry.Token, &encryptedToken)
	}

	return nil
}
```

This also has the benefit of validating *before* applying the update (avoiding in-memory mutations that are immediately discarded), and of skipping the `crypto.Encrypt` call when the username alone is already invalid.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Last reviewed commit: d44e145</sub>

> Greptile also left **1 inline comment** on this PR.

<sub>(3/5) Reply to the agent's comments like "Can you suggest a fix for this @greptileai?" or ask follow-up questions!</sub>

### CODE REVIEW

File: backend/internal/services/container_registry_service.go [hunk 1]
```diff
@@ -260,15 +254,18 @@ func (s *ContainerRegistryService) updateECRRegistryFieldsInternal(registry *mod
...
 
-	encryptedToken, err := crypto.Encrypt(*req.Token)
-	if err != nil {
-		return fmt.Errorf("failed to encrypt token: %w", err)
+	if strings.TrimSpace(registry.Username) == "" {
+		return &models.ValidationError{Field: "username", Message: "Username is required"}
 	}
-	utils.UpdateIfChanged(&registry.Token, encryptedToken)
+
 	return nil
```
[the remark this answers is not in the document]
Comment (paulonnn) [reply backend/internal/services/container_registry_service.go:269]: @kmendell i dont think this is a problem at all, since the username field should be required anyway even before this implementation, am i correct?

[the remark this answers is not in the document]
Comment (kmendell) [reply backend/internal/services/container_registry_service.go:269]: Yes it shouldn't matter as username should have been required.
